### PR TITLE
policy: Fix Distillery Tests

### PR DESCRIPTION
**Pull request by nathanjsweet**
See commits for details.

### CODE REVIEW

File: pkg/policy/distillery_test.go [hunk 1]
```diff
@@ -1624,12 +1625,12 @@ var (
 					{
 						Port:     "64",
 						EndPort:  127,
-						Protocol: api.ProtoAny,
```
Comment (squeed) [inline pkg/policy/distillery_test.go:1627]: Why are we changing the test data here?

File: pkg/policy/distillery_test.go [hunk 2]
```diff
@@ -1663,72 +1669,83 @@ func Test_EnsureDeniesPrecedeAllows(t *testing.T) {
...
-		rules  api.Rules
-		result MapState
+		test     string
+		rules    api.Rules
+		expected MapState
 	}{
 		{"deny_world_no_labels", api.Rules{ruleL3DenyWorld, ruleL3AllowWorldIP}, newMapState(map[Key]MapStateEntry{
-			mapKeyL3WorldIngress: mapEntryDeny,
-			mapKeyL3WorldEgress:  mapEntryDeny,
+			mapKeyL3WorldIngress:  mapEntryDeny,
```
Comment (squeed) [inline pkg/policy/distillery_test.go:1663]: Why are there now 2 keys for the same direction? Shouldn't a single wildcard be sufficient?

File: pkg/policy/distillery_test.go [hunk 3]
```diff
@@ -1663,72 +1669,83 @@ func Test_EnsureDeniesPrecedeAllows(t *testing.T) {
...
 		})}, {"deny_world_with_labels", api.Rules{ruleL3DenyWorldWithLabels, ruleL3AllowWorldIP}, newMapState(map[Key]MapStateEntry{
-			mapKeyL3WorldIngress: mapEntryWorldDenyWithLabels,
-			mapKeyL3WorldEgress:  mapEntryWorldDenyWithLabels,
-		})}, {"deny_one_ip_with_a_larger_subnet", api.Rules{ruleL3DenySubnet, ruleL3AllowWorldIP}, newMapState(map[Key]MapStateEntry{
+			mapKeyL3WorldIngress:  mapEntryWorldDenyWithLabels,
+			mapKeyL3WorldEgress:   mapEntryWorldDenyWithLabels,
 			mapKeyL3SubnetIngress: mapEntryDeny,
 			mapKeyL3SubnetEgress:  mapEntryDeny,
+		})}, {"deny_one_ip_with_a_larger_subnet", api.Rules{ruleL3DenySubnet, ruleL3AllowWorldIP}, newMapState(map[Key]MapStateEntry{
+			mapKeyL3SubnetIngress:        mapEntryDeny,
```
Comment (squeed) [inline pkg/policy/distillery_test.go:1673]: same question here -- why are there new mapstate entries here?

**Review by chancez — APPROVED** [review]

**Comment by squeed** [comment]
It seems like the only logic difference is that we now check for supersets w.r.t owners. Is that correct?

File: pkg/maps/policymap/policymap.go [hunk 4]
```diff
@@ -117,6 +117,11 @@ func (k *PolicyKey) GetDestPort() uint16 {
 	return byteorder.NetworkToHost16(k.DestPortNetwork)
 }
 
+// GetPortMask returns the port mask of the key
+func (k *PolicyKey) GetPortMask() uint16 {
+	return 0xffff
```
Comment (jrajahalme) [inline pkg/maps/policymap/policymap.go:122]: Should this be conditional on `DestPortNetwork` being 0?

Comment (nathanjsweet) [reply pkg/maps/policymap/policymap.go:122]: Eventually. The pull request that finally introduces port ranges will add logic here.

File: pkg/policy/distillery_test.go (hunk 1 above)
Comment (nathanjsweet) [reply pkg/policy/distillery_test.go:1627]: To reduce the overall test data before. The test was actually failing before and required a lot more entries than was in the "expected" field. Changing this to TCP reduces the overall test data without sacrificing a comprehensive test of the logic.

File: pkg/policy/distillery_test.go (hunk 2 above)
Comment (nathanjsweet) [reply pkg/policy/distillery_test.go:1663]: Other CIDR identities are automatically selected by the reserved:World identity by the policy repository. This test was silently failing before, nothing in the port range work caused this change. We can and should address/understand these redundancies, but not in this PR.

File: pkg/policy/distillery_test.go (hunk 3 above)
Comment (nathanjsweet) [reply pkg/policy/distillery_test.go:1673]: I'm not sure what change caused this, but I'm confident it was not the port range changes.

**Comment by nathanjsweet** [comment]
> It seems like the only logic difference is that we now check for supersets w.r.t owners. Is that correct?

I removed the ownership changes, they can wait for another PR (it was a simple optimization). The only logic changes in this PR now is the inverted port mask fix and the deletion logic fix.

**Review by jrajahalme — APPROVED** [review]

**Review by squeed — APPROVED** [review]

File: pkg/policy/types/types.go [hunk 5]
```diff
@@ -121,26 +121,10 @@ func (k Key) PortIsEqual(c Key) bool {
 		k.InvertedPortMask == c.InvertedPortMask
 }
 
-// maskPrefixMap is the map of all the possible portMasks and their
-// associated prefix lengths.
-var maskPrefixMap = map[uint16]uint{
```
Comment (gentoo-root) [inline pkg/policy/types/types.go:126]: The old code returned the number of leading zeroes, the new code switched to trailing. Is it intended? Was the old code wrong?

File: pkg/policy/mapstate.go [hunk 6]
```diff
@@ -861,21 +867,25 @@ func (ms *mapState) denyPreferredInsertWithChanges(newKey Key, newEntry MapState
...
-				updates = append(updates, MapChange{
+				deletes = append(deletes, MapChange{
 					Key: k,
 				})
+
 			}
 			return true
 		})
+		for _, delete := range deletes {
+			if !delete.Add {
```
Comment (gentoo-root) [inline pkg/policy/mapstate.go:878]: Seems like the `!delete.Add` and `update.Add` checks are not strictly needed.

**Review by gentoo-root — APPROVED** [review]

File: pkg/policy/mapstate.go (hunk 6 above)
Comment (nathanjsweet) [reply pkg/policy/mapstate.go:878]: The are not.

File: pkg/policy/types/types.go (hunk 5 above)
Comment (nathanjsweet) [reply pkg/policy/types/types.go:126]: The logic is correct, the name is poor.

**Pull request merged**